include/parsers/qasm3_parser/Parser.hpp

```diff
@@ -0,0 +1,194 @@
+ * more information.
+ */
+
+#pragma once
+
+#include "Definitions.hpp"
+#include "Scanner.hpp"
+#include "Statement.hpp"
+#include "StdGates.hpp"
+#include "operations/CompoundOperation.hpp"
+#include "operations/NonUnitaryOperation.hpp"
+#include "operations/StandardOperation.hpp"
+
+#include <cmath>
+#include <iostream>
+#include <regex>
+#include <set>
+#include <sstream>
+#include <stdexcept>
+#include <utility>
+#include <vector>
+
+namespace qasm3 {
+class Parser {
+private:
+  struct ScannerState {
+  private:
+    std::unique_ptr<std::istream> is;
+
+  public:
+    Token t{0, 0};
+    Token next{0, 0};
+    std::unique_ptr<Scanner> scanner;
+    std::optional<std::string> filename;
+    bool isImplicitInclude;
+
+    bool scan() {
+      t = next;
+      next = scanner->next();
+
+      return t.kind != Token::Kind::Eof;
+    }
+
+    explicit ScannerState(
+        std::istream& in,
+        std::optional<std::string> debugFilename = std::nullopt,
+        bool implicitInclude = false)
+        : scanner(std::make_unique<Scanner>(in)),
+          filename(std::move(debugFilename)),
+          isImplicitInclude(implicitInclude) {
+      scan();
+    }
+
+    explicit ScannerState(
+        std::unique_ptr<std::istream> in,
+        std::optional<std::string> debugFilename = std::nullopt,
+        bool implicitInclude = false)
+        : is(std::move(in)), scanner(std::make_unique<Scanner>(*is)),
+          filename(std::move(debugFilename)),
+          isImplicitInclude(implicitInclude) {
+      scan();
+    }
+  };
+
+  std::stack<ScannerState> scanner{};
+  std::shared_ptr<DebugInfo> includeDebugInfo{nullptr};
+
+  [[noreturn]] static void error(const Token& token, const std::string& msg) {
+    std::cerr << "Error at line " << token.line << ", column " << token.col
+              << ": " << msg << '\n';
+    throw std::runtime_error("Parser error");
+  }
+
+  static void warn(const Token& token, const std::string& msg) {
+    std::cerr << "Warning at line " << token.line << ", column " << token.col
+              << ": " << msg << '\n';
+  }
+
+  [[nodiscard]] inline Token current() const {
+    if (scanner.empty()) {
+      throw std::runtime_error("No scanner available");
+    }
+    return scanner.top().t;
+  }
+
+  [[nodiscard]] inline Token peek() const {
+    if (scanner.empty()) {
+      throw std::runtime_error("No scanner available");
+    }
+    return scanner.top().next;
+  }
+
+  Token expect(const Token::Kind& expected,
+               std::optional<std::string> context = std::nullopt) {
+    if (current().kind != expected) {
+      std::string message = "Expected '" + Token::kindToString(expected) +
+                            "', got '" + Token::kindToString(current().kind) +
+                            "'";
+      if (context.has_value()) {
+        message += " " + context.value();
+      }
+      error(current(), message);
+    }
+
+    Token const token = current();
+    scan();
+    return token;
```

Comment:
```suggestion
    auto token = current();
    scan();
    return token;
```
Should resolve the clang-tidy warning